pkg/apiserver/apic.go

```diff
@@ -0,0 +1,231 @@
+		alertToPush:     make(chan []*models.Alert),
+		dbClient:        dbClient,
+		pullInterval:    pullInterval,
+		pushInterval:    pushInterval,
+		metricsInterval: metricsInterval,
+		mu:              sync.Mutex{},
+		startup:         true,
+		credentials:     config.Credentials,
+		pullTomb:        tomb.Tomb{},
+		pushTomb:        tomb.Tomb{},
+		metricsTomb:     tomb.Tomb{},
+	}, nil
+}
+
+func (a *apic) Push() error {
+	defer types.CatchPanic("apil/pushToAPIC")
+
+	var cache []*models.Alert
+	ticker := time.NewTicker(a.pushInterval)
+
+	for {
+		select {
+		case <-a.pushTomb.Dying(): // if one apic routine is dying, do we kill the others?
+			a.pullTomb.Kill(nil)
+			a.metricsTomb.Kill(nil)
+			log.Infof("push tomb is dying, sending cache (%d elements) before exiting", len(cache))
+			err := a.Send(cache)
+			return err
+		case <-ticker.C:
+			// flush
+			a.mu.Lock()
+			err := a.Send(cache)
+			if err != nil {
+				return err
+			}
+			cache = make([]*models.Alert, 0)
+			a.mu.Unlock()
+		case alerts := <-a.alertToPush:
+			a.mu.Lock()
+			cache = append(cache, alerts...)
+			a.mu.Unlock()
+		}
+	}
+}
+
+func (a *apic) Send(cache []*models.Alert) error {
+	_, _, err := a.apiClient.Alerts.Add(context.Background(), cache)
+	return err
+}
+
+func (a *apic) Pull() error {
+	defer types.CatchPanic("apil/pullFromAPIC")
+
+	ticker := time.NewTicker(a.pullInterval)
+	for {
+		select {
+		case <-ticker.C:
+			data, _, err := a.apiClient.Decisions.GetStream(context.Background(), a.startup)
+			if err != nil {
+				return errors.Wrap(err, "pull top")
+			}
+			if a.startup {
+				a.startup = false
+			}
+			for _, decision := range data.New {
+				alertCreated, err := a.dbClient.Ent.Alert.
+					Create().
+					SetScenario(*decision.Scenario).
+					SetSourceIp(*decision.Value).
+					//SetSourceAsNumber(alert["as_num"]).
+					//SetSourceAsName(alert["as_org"]).
+					//SetSourceCountry(alert["country"]).
+					Save(a.dbClient.CTX)
+				if err != nil {
+					log.Fatalf("unable to create alors from topX: %s", err)
+				}
+
+				duration, err := time.ParseDuration(*decision.Duration)
+				if err != nil {
+					log.Fatalf("unable to parse decision duration '%s': %s", *decision.Duration, err)
+				}
+				startIP, endIP, err := controllers.GetIpsFromIpRange(*decision.Value)
+				if err != nil {
+					log.Fatalf("failed querying alerts: Range %v is not valid", *decision.Value)
+				}
+
+				_, err = a.dbClient.Ent.Decision.Create().
+					SetUntil(time.Now().Add(duration)).
+					SetScenario(*decision.Scenario).
+					SetType(*decision.Type).
+					SetStartIP(startIP).
+					SetEndIP(endIP).
+					SetValue(*decision.Value).
+					SetScope(*decision.Scope).
+					SetOwner(alertCreated).Save(a.dbClient.CTX)
+				if err != nil {
+					log.Fatalf("failed creating decision from top: %v", err)
+				}
+			}
+			log.Printf("pull top: added %d entries", len(data.New))
+
+			var filter map[string][]string
+			for _, decision := range data.Deleted {
```

Comment:
I think we want to process `Deleted` before `Added` in case we would have a deleted and added decision on a similar scope within the same pull ?